### PR TITLE
Modal dialogs

### PR DESCRIPTION
**Pull request by ZanderBrown**
Mode & Admin were showing some odd behaviours (at least on mutter/GNOME but I suspect others are affected) largely because they were being modal to an unset parent

This is slightly WIP as FileDialogs still aren't working properly but I'm 99% sure that's a problem in `QGtk3FileDialogHelper` (aka upstream)

### CODE REVIEW

**Comment by carlosperate** [comment]
Would this change affect the behaviour described in https://github.com/mu-editor/mu/issues/498?
I'm wondering if having the two windows (mode selection and swicth to microbit/adafruit mode) could cause some weird scenario where one blocks the other.

File: mu/interface/dialogs.py [hunk 1]
```diff
@@ -48,6 +48,9 @@ class ModeSelector(QDialog):
     Defines a UI for selecting the mode for Mu.
     """
 
+    def __init__(self, parent):
+        super().__init__(parent)
+
```
Comment (ntoll) [inline mu/interface/dialogs.py]: Surely these redefinitions of `__init__` are surplus to requirements..? If they didn't exist the `__init__` on the base class would be called anyway (and that's simply what you're manually doing here). Right..? Or am I missing something here?

File: tests/interface/test_dialogs.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ def test_ModeSelector_setup():
     current_mode = 'python'
     mock_item = mock.MagicMock()
     with mock.patch('mu.interface.dialogs.ModeItem', mock_item):
-        ms = mu.interface.dialogs.ModeSelector()
+        ms = mu.interface.dialogs.ModeSelector(None)
```
Comment (ntoll) [inline tests/interface/test_dialogs.py]: Why is `None` needed here..?

Also, CI complains because the tests don't pass. :-(

**Comment by ntoll** [comment]
Hmm... if we can improve the behaviour of modal dialogs then this is a good thing (tm). But I'm confused (see comments) about some of the changes (I'm probably missing something here). Also, the tests aren't passing.

When the above is addressed, happy to merge this. :-)

File: mu/interface/dialogs.py (hunk 1 above)
Comment (ZanderBrown) [reply mu/interface/dialogs.py]: Yes the base `__init__` would be called but with `parent=None`. Qt tries to be clever and makes the dialog 'application modal' meaning it puts the dialog to the top whenever another window is raised above it. End result is magic disappearing then reappearing dialogs.

By setting the parent here Qt can set the correct Window Manager hints so the dialog is modal to the parent properly

I'm not sure I am explaining this well

File: tests/interface/test_dialogs.py (hunk 2 above)
Comment (ZanderBrown) [reply tests/interface/test_dialogs.py]: With hindsight `__init__` should have `parent=None`

**Comment by ntoll** [comment]
@ZanderBrown the tests still don't pass... see the output from Travis, specifically:

```
>           mock_admin_display.assert_called_once_with()
tests/interface/test_main.py:1132: 
```

**Comment by ZanderBrown** [comment]
:man_facepalming: 

Sorry haven't been running tests locally because they require QtChart which i don't currently have after debugging the RPM...

**Comment by ntoll** [comment]
Great stuff..! Tests passing (but OSX build is knackered, I'm fixing it tomorrow).

**Pull request merged**